### PR TITLE
feat: CLI `config` not supporting nested properties correctly

### PR DESCRIPTION
**Pull request by BlueWinds**
- Closes UNIFY-1028

### User facing changelog
Can now pass in testing-type specific config options via `--config` without nesting JSON. 

### Additional details
With the addition of more testing type specific config options, and especially with the requirement that `specPattern` be configured per testing type, using `--config` via command line was a bit clunky, requiring a user to nest the config value.

#### Before

```sh
$ cd packages/driver
$ node ../../scripts/cypress run '--config={"e2e": {"specPattern": "**/*.spec.jsx"}}'
# Works!

$ node ../../scripts/cypress run '--config={"specPattern": "**/*.spec.jsx"}'
`specPattern` is not a valid configuration option

https://on.cypress.io/configuration

`specPattern` is not a valid configuration option
```

#### After

```sh
$ node ../../scripts/cypress run '--config={"e2e": {"specPattern": "**/*.spec.jsx"}}'
# Still works!

$ node ../../scripts/cypress run '--config={"specPattern": "**/*.spec.jsx", "baseUrl": "https://example.com"}'
# Now also works
```

This example resolves to the the config of
```js
{
  component: {
    specPattern: "**/*.spec.jsx" // specPattern is a valid config for both component and e2e testing, so it is included in both.
  },
  e2e: {
    baseUrl: "https://example.com", // baseUrl is valid only for e2e, so it is included only here.
    specPattern: "**/*.spec.jsx"
  }
}
```

### PR Tasks
- [x] Have tests been added/updated?
- [N/A] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)?
- [N/A] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
- [N/A] Have new configuration options been added to the [`cypress.schema.json`](https://github.com/cypress-io/cypress/blob/develop/cli/schema/cypress.schema.json)?

### CODE REVIEW

**Comment by BlueWinds** [comment]
A number of tests are failing here on this branch, but I'm seeing the same failures in 10.0-release, and even develop when running locally - it feels like something changed with regards to cookies recently, perhaps a browser version upgrade?

Going to put this as ready for review while I look into it, since I suspect the solution is unrelated to the changes in here.

**Comment by emilyrohrbough** [comment]
Without looking at the changes, could I run something like:
```
--e2e --config= "specPattern":["*/test2.coffee","*/test1.js"]}
```

File: packages/server/test/unit/util/args_spec.js [hunk 1]
```diff
@@ -595,6 +595,20 @@ describe('lib/util/args', () => {
         config: {},
       })
     })
+
+    it('moves testing-type specific config options', function () {
```
Comment (lmiller1990) [inline packages/server/test/unit/util/args_spec.js:599]: Cool, looks good. Do we have a pre-existing pattern for if they pass invalid config, eg `--config={blablah=false}`, or we just ignore these? It looks like we will inherit the same behavior with the hoisted keys, just wanted to clarify.

**Review by lmiller1990 — APPROVED** [review]
One question. Also, rebased it, so we should be good to get another :heavy_check_mark: and merge.

**Comment by cowboy** [comment]
> I want to clarify: we use JSON for `--config`, but the actual config is no longer JSON, but a JS module. Do you know if this has been discussed (stick w/ JSON on config, even if we are not using JSON for the `cypress.config.js`. What about if they try to pass a key that's a function, like `devServer`? I wonder if we should error.
> 
> I'm guessing this is something @cypress-io/dx has considered previously when thinking about `cypress.config.js`?

We did discuss the `--config` option, and what I remember from that conversation (I couldn't find any documentation) was that all options specified via `--config` override the _resolved_ config value, so that the user never needs to specify the `e2e` or ` component` objects. It was my understanding that this is the 9.x behavior, and we were going to keep it in 10.0.

Also, I believe we decided that function options would _not_ be able to be specified via `--config`.

Does this make sense? Are there any issues with this?

**Comment by cowboy** [comment]
> ```
> --e2e --config='{"specPattern":["*/test2.coffee","*/test1.js"]}'
> ```

Not sure if this is relevant, but Windows cmd shell only understands `"` as a string delimiter, in case we're testing this there.

**Comment by BlueWinds** [comment]
> We did discuss the --config option, and what I remember from that conversation (I couldn't find any documentation) was that all options specified via --config override the resolved config value, so that the user never needs to specify the e2e or component objects. It was my understanding that this is the 9.x behavior, and we were going to keep it in 10.0.

> Also, I believe we decided that function options would not be able to be specified via --config.

> Does this make sense? Are there any issues with this?

Makes sense to me, and matches what was implemented in the PR, thanks for explaining. Just wanted to double check to make sure this was desired and intended.

> > ```
> > --e2e --config='{"specPattern":["*/test2.coffee","*/test1.js"]}'
> > ```
> 
> Not sure if this is relevant, but Windows cmd shell only understands `"` as a string delimiter, in case we're testing this there.

No tests / docs using single quotes, provided here only as an example for discussion in the PR.

**Review by lmiller1990 — APPROVED** [review]
thanks for clarification @cowboy, seems we are all good.

**Review by brian-mann — CHANGES_REQUESTED** [review]
This was something we discussed during the design of the CLI changes as per 10.0

In fact there is not supposed to be support for passing nested `--config` values for e2e + component. They are supposed to be passed **flattened** for the testing type that you want applied.

So passing `--config specPattern='...'` is only applied **to the testing type you eventually go into** and is applied to both testing types. You don't pass different configuration to different testing types.

It this is something we'd like to revisit I'm all ears - but it was intentionally designed this way. It's possible the existing implementation was not working correctly, but the aforementioned behavior was the desired behavior.

### Edit
Sorry I just reread the comments throughout this PR and clearly others have chimed in with the same information. @BlueWinds what's not clear from this PR is what specific behavior this is fixing. When I first read this I saw the PR description regarding the `config` not supporting nesting and assumed it was attempting to pass specific testing type config. What I'm looking for clarity on is what specific **existing** behavior is and what this PR is changing to the **desired** behavior. A simple before/after.

**Comment by lmiller1990** [comment]
I just tested this out and as @brian-mann points out, the flat/merging stuff actually already works as-expected in `10.0-release`. Maybe this does not actually fix anything?

**Comment by BlueWinds** [comment]
> Sorry I just reread the comments throughout this PR and clearly others have chimed in with the same information. @BlueWinds what's not clear from this PR is what specific behavior this is fixing. When I first read this I saw the PR description regarding the `config` not supporting nesting and assumed it was attempting to pass specific testing type config. What I'm looking for clarity on is what specific **existing** behavior is and what this PR is changing to the **desired** behavior. A simple before/after.

Latchaln put this fairly clearly, so I'll refer you to his comment for a before / after.

Currently in 10.0, *most* config values already work as you describe, getting included in whichever config type the user selects. The only ones that don't are those explicitly not available as root config values - eg, those in [breakingRootOptions](https://github.com/cypress-io/cypress/blob/issue-UNIFY-1028-cli-nested-config/packages/config/lib/options.ts#L532): `supportFile`, `specPattern`, `excludeSpecPattern` and `baseUrl`. These are not valid as 'root' config options, and so the CLI rejects them.

This PR aims to allow those breakingRootOptions to be applied to the appropriate testing type when passed in via --config. I'll update the PR description once I address @lmiller1990's findings above.

**Comment by BlueWinds** [comment]
> I tested this branch a bit more. This one isn't working for me unless I explicitly include `--e2e` or `--component`. I'm not sure if that's related, and it only seems to be for nested only config options (like `specPattern`).

It was checking to see what testing type the user had actually passed in, which, upon reflection, is not the correct action - we *know* what testingType `baseUrl` applies to. I've updated the PR so that it nests invalid root configs directly to the testing types they apply to - it will also now work in open mode when you switch between testing types, applying CLI args to the appropriate one.

**Comment by BlueWinds** [comment]
Updated PR with clear before / after examples.

**Review by lmiller1990 — APPROVED** [review]
Cool, seems good to me.

**Comment by flotwig** [comment]
Seems like a legit failure in server-unit-tests: https://app.circleci.com/pipelines/github/cypress-io/cypress/32422/workflows/962760ee-0761-4d37-b915-8babc3bf8e5d/jobs/1284475

**Comment by BlueWinds** [comment]
> Seems like a legit failure in server-unit-tests: https://app.circleci.com/pipelines/github/cypress-io/cypress/32422/workflows/962760ee-0761-4d37-b915-8babc3bf8e5d/jobs/1284475

Fixed.

**Pull request merged**